hack around Xlib deadlock

Be-ing opened the pull request:
adapted code from https://gitlab.freedesktop.org/xorg/lib/libx11/issues/25#note_17248
https://bugs.launchpad.net/mixxx/+bug/1805559

I have not had time to extensively test this yet.

Be-ing committed:
hack around Xlib deadlock

adapted code from https://gitlab.freedesktop.org/xorg/lib/libx11/issues/25#note_17248
https://bugs.launchpad.net/mixxx/+bug/1805559

Be-ing commented:
Oops this breaks running with `-platform wayland`...

Be-ing committed:
only hack around Xlib bug when running as an X client

Otherwise calling Xlib functions segfaults when running as a
Wayland client

uklotzde commented:
I've cherry-pricked the commits for testing. But since I didn't experience the issue I can only test for any unwanted side effects.

Running Mixxx with QtWayland never really worked, even without this patch. The UI is incomplete and Mixxx is unresponsive, so I won't test this use case.

Be-ing commented:
Yes the QPA Wayland backend has many issues that make it not ready for production. Just at first glance, high DPI scaling is somehow handled differently (a scale factor of 1.5 looks the same as 3.0 with X??) and my theme does not apply. I just wanted to make sure this didn't break running with the QPA Wayland backend any more than Qt is broken with it. I would not be surprised if it becomes the default in the future.

uklotzde commented on src/mixxx.cpp:
```diff
@@ -81,10 +81,41 @@
 #include "preferences/dialog/dlgprefmodplug.h"
 #endif
 
+#ifdef Q_OS_LINUX
```
Please replace all occurrences of `#ifdef Q_OS_LINUX` in this patch with `#if defined(Q_WS_X11)` that we already used for this purpose in other parts of Mixxx.

Be-ing replied:
I can't find any documentation for Q_WS_X11 for Qt5. Conceptually it doesn't make sense. The window system is not determined at compile, but at runtime. I think Q_OS_LINUX is correct.

uklotzde replied:
What happens if you neither have X11 nor its development headers installed? This might not happen soon, but it will happen. We use Qt as an abstraction and it doesn't matter, but here the abstractions leaks through at compile time.

Be-ing replied:
I think we can defer that until it becomes a real problem. If Qt doesn't help us when that comes, we'll have to detect the presence of X headers in our build system.

uklotzde commented on src/mixxx.cpp:
```diff
@@ -81,10 +81,41 @@
+typedef Bool (*WireToErrorType)(Display*, XErrorEvent*, xError*);
+
+const int NUM_HANDLERS = 256;
+WireToErrorType __oldHandlers[NUM_HANDLERS] = {0};
```
I still don't get the magic behind this array. Which unknown side effect sets the elements to a value other than 0? Currently the error interceptor does nothing except ignoring and dropping all incoming extension errors. This can be implemented in a much simpler way.

Be-ing replied:
Honestly, neither do I, nor do I understand the magic number 256. I've never worked with X11 code before, much less understand X's internals. I just copied and pasted from https://bugs.freedesktop.org/show_bug.cgi?id=59361#c4 and got it to build.

uklotzde replied:
256 because the error code is an unsigned char with values ranging from 0 to 255. But all slots are initialized with 0 (= `nullptr`) and are never updated.

Be-ing committed:
use #if defined rather than #ifdef

uklotzde commented on src/mixxx.cpp:
```diff
@@ -81,10 +81,41 @@
 #include "preferences/dialog/dlgprefmodplug.h"
 #endif
 
+#ifdef Q_OS_LINUX
```
This occurrence also needs to be replaced.

uklotzde commented:
Missing dependency in .travis.yml!

uklotzde commented:
Next failure during linking:
```
/usr/bin/ld: lin64_build/mixxx.o: in function `MixxxMainWindow::MixxxMainWindow(QApplication*, CmdlineArgs const&)':
/home/uk/Projects/Mixxx/mixxx/src/mixxx.cpp:187: undefined reference to `qt_x11_wait_for_window_manager(QWidget*)'
collect2: error: ld returned 1 exit status
```

Be-ing committed:
fix #include

Be-ing committed:
remove call to qt_x11_wait_for_window_manager

This was previously guarded by an #ifdef Q_WS_X11, but switching
to Q_OS_LINUX caused a link error:

/usr/bin/ld: lin64_build/mixxx.o: in function `MixxxMainWindow::MixxxMainWindow(QApplication*, CmdlineArgs const&)':
/home/uk/Projects/Mixxx/mixxx/src/mixxx.cpp:187: undefined reference to `qt_x11_wait_for_window_manager(QWidget*)'

It turns out that Q_WS_X11 is no longer defined with Qt5, yet it
does not seem that building Mixxx without the call to
qt_x11_wait_for_window_manager has caused any issues, so I presume
it is safe to remove it now. This was the only place in Mixxx
where Q_WS_X11 was referenced.

uklotzde commented:
libqt5x11extras5-dev is still missing in .travis.yml to fix the CI build

Be-ing committed:
add libqt5x11extras5-dev to Travis dependencies

uklotzde commented:
Using the unmodified patch is feasible since we have no clue how the code actually is intended to work. As far as I understand it suppresses errors to avoid getting stuck at a subsequent lock.

If this workaround is effective (what I'm not able to validate) I have no objections merging it. LGTM.

Be-ing commented:
I experienced no crash after playing for 1.5 hours. Of course that doesn't necessarily mean the problem is solved, but I'll keep testing.
